src/mlpack/core/metrics/non_maximal_supression.hpp

```diff
@@ -0,0 +1,75 @@
+/**
+ * @file non_maximal_supression.hpp
+ * @author Kartik Dutt
+ *
+ * Definition of Non Maximal Supression metric.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_METRICS_NMS_HPP
+#define MLPACK_CORE_METRICS_NMS_HPP
+
+namespace mlpack {
+namespace metric {
+
+/**
+ * Definition of Non Maximal Supression.
+ *
+ * Performs non-maximal suppression (NMS) on the boxes according to their
+ * intersection-over-union (IoU). NMS iteratively removes lower scoring boxes
+ * which have an IoU greater than threshold with another high scoring box.
+ *
+ * For bounding box representation there are two common representation
+ * either as coordinates i.e. each value in vector represents a
+ * coordinate in the format x0, y0, x1, y1 where x0, y0 represent the
+ * lower left coordinate and x1, y1, represent upper right coordinate.
+ * 
+ * Second representation follows the following representation : x0, y0, h, w.
+ * Where x0 and y0 are bottom left bounding box coordinates and h, w are
+ * height and width of the bounding box.
+ *
+ * @tparam UseCoordinates Toggles between the two representation of bounding box.
+ *                        If true, each value in vector represents a coordinate 
+ *                        in the formate x0, y0, x1, y1. Else the bounding box is
+ *                        represented as x0, y0, h, w.
+ */
+template<
+    bool UseCoordinates = false
+>
+class NMS
+{
+ public:
+  //! Default constructor required to satisfy the Metric policy.
+  NMS() { /* Nothing to do here. */ }
+
+  /**
+   * Performs non-maximal suppression.
+   */
+  template<
+      typename BoundingBoxesType,
+      typename ConfidenceScoreType,
+      typename OutputType
+  >
+  static void Evaluate(const BoundingBoxesType& boundingBoxes,
+                       const ConfidenceScoreType& confidenceScores,
+                       OutputType& selectedBoundingBoxes,
```

Comment:
What about return **indices** as `selectedIndices`, not bound boxes? I think it would be better for the usability... and I'm sorry about that I didn't mention it in the previous discussion.

Reply:
PyTorch returns indices but I wasn't sure if that would be the way to go because the user would have to selected bounding boxes, This reduces their effort. Let me know what you think.

Reply:
For some reason, I thought users might need to get indices not real coordinates. By just giving indices as a output, we can don't care the case and users can utilize both the indices and coordinates.

Reply:
Ahh, makes sense. I'll make the changes and adapt the tests.